src/license_guc.c

```diff
@@ -107,279 +101,99 @@ ts_license_enable_module_loading(void)
-	 * re-parse in the assign hook
-	 */
-	extra = NULL;
-#endif
-
-	module_can_start = DirectFunctionCall2(tsl_validate_license_fn,
-										   CStringGetDatum(*newval),
-										   PointerGetDatum(extra));
-
-	return DatumGetBool(module_can_start);
-}
-
-#ifdef WIN32
-static void *
-revalidate_license(const char *newval)
-{
-	void *extra = NULL;
-	void **extra_p = &extra;
-
-	Assert(extra == NULL);
-
-	/*
-	 * Due to windows issues we cannot use the extra parameter. Instead
-	 * re-call the validation function, if we reach this point the license
-	 * must be valid so the function cannot fail cannot fail.
-	 */
-	DirectFunctionCall2(tsl_validate_license_fn, CStringGetDatum(newval), PointerGetDatum(extra_p));
-	return extra;
+	tsl_init_fn = function;
+	tsl_handle = handle;
+	return true;
 }
-#endif
 
-/*
- * `ts_license_on_assign`
- * Links the cross-module function pointers.
- */
-void
-ts_license_on_assign(const char *newval, void *extra)
+static void
+tsl_module_init(void)
 {
-	if (!can_load)
-		return;
-
-	Assert(newval != NULL);
-	Assert(TS_LICENSE_TYPE_IS_VALID(newval));
-	if (TS_LICENSE_IS_APACHE_ONLY(newval))
-	{
-		Assert(current_license_can_downgrade_to_apache());
-		Assert(extra == NULL);
-		if (ts_cm_functions->module_shutdown_hook != NULL)
-			ts_cm_functions->module_shutdown_hook();
-		return;
-	}
-
 	Assert(tsl_handle != NULL);
-	Assert(tsl_startup_fn != NULL);
-	DirectFunctionCall1(tsl_startup_fn, CharGetDatum(0));
-#ifdef WIN32
-	Assert(extra == NULL);
-	extra = revalidate_license(newval);
-#endif
-	Assert(extra != NULL);
-	ts_cm_functions->tsl_license_on_assign(newval, extra);
+	Assert(tsl_init_fn != NULL);
+	DirectFunctionCall1(tsl_init_fn, CharGetDatum(0));
 }
 
-/* Module Functions */
-
 /*
- * Load the module, but do not start it.
- * If this function succeeds, `tsl_handle`, and
- * `tsl_startup_fn` will be set, but no `TS_FN` functions will
- * change. `tsl_startup_fn` can be used to enable the submodule
- * versions of `TS_FN`s.
- *
- * This function is idempotent, and will _not_ reload the module
- * if called multiple times.
- *
- * returns:
- *   a function pointer to `tsl_license_update_check` or NULL if the
- *   loading fails
+ * Check hook function set by license guc.
  *
+ * Used to validate license string before the assign hook
+ * ts_license_guc_assign_hook() call.
  */
-static bool load_tsl(void);
-
-static bool
-load_tsl(void)
+bool
+ts_license_guc_check_hook(char **newval, void **extra, GucSource source)
 {
-	char soname[MAX_SO_NAME_LEN] = { 0 };
+	LicenseType type = license_type_of(*newval);
 
-	if (tsl_handle != NULL)
+	/* Allow setting a license only if is is set from postgresql.conf
+	 * or the server command line */
+	switch (type)
 	{
-		Assert(tsl_startup_fn != NULL);
+		case LICENSE_APACHE:
+		case LICENSE_TSL:
+			if (source == PGC_S_FILE || source == PGC_S_ARGV || source == PGC_S_DEFAULT)
+				break;
+			GUC_check_errdetail("Cannot change a license in a running session.");
```

Comment:
Can you provide detail from which value to which the change requested?

Reply:
Not sure this is relevant here because it can't change any license runtime

Reply:
I agree that it is not relevant. Is there a test, which runs into this error? It will be more clear if it is already good enough by looking to the actual scenarios when it happens.

Reply:
Yes, it is tested in test/sql/license.sql

Reply:
@pmwkaa Is there a test, which gets this error?

Reply:
https://github.com/timescale/timescaledb/pull/2435/files#diff-a1388d76a54b03a1aea1f9c9b4fe3749R20

Reply:
Thank you for the link! Looks good!